### PR TITLE
[Spec Decode] Enable fused draft metadata updates for sparse MLA backends under DCP

### LINKED ISSUE #54369: [Perf][ROCm] Sparse-MLA/kpool backends fall back to rebuilding attention metadata per draft step, capping useful MTP depth at k=4
### Summary

On ROCm with GLM-5.3-Flash, vLLM logs at startup:

```
Fused multi-step draft decode is not supported by attention backend(s)
DEEPSEEK_V32_INDEXER, KPOOL_TAIL, ROCM_AITER_MLA_SPARSE;
falling back to rebuilding attention metadata between draft steps.
```

Because the metadata is rebuilt between every draft step, the per-step cost of MTP speculation is high enough that deeper speculation stops paying for itself well before the draft quality runs out. Measurements below quantify the gap.

### Measurements

8x MI350X (gfx950), ROCm 7.2.3, TP8, GLM-5.3-Flash-FP8, `--max-model-len 1048576`, 2 000-token context, 256 output tokens, unique per-request prefixes (nothing hits the prefix cache), warmup round then median of 3-4 rounds.

| `num_speculative_tokens` | acceptance length | draft acceptance rate | 1 user | 16 users | 64 users |
|---|---|---|---|---|---|
| 2 | 2.57 | 78.4 % | 155 tok/s | 1 432 | 3 012 |
| **4** | **3.36** | 58.9 % | **173** | **1 499** | **3 055** |
| 6 | 3.76 | 46.0 % | 175 | 1 367 | 2 874 |

The accepted length keeps climbing (2.57 → 3.36 → 3.76), i.e. the draft head still has useful signal at k=6. But throughput peaks at k=4 and *falls* at k=6, which is what you would expect when each additional draft step carries a fixed metadata-rebuild cost rather than being fused.

Note this model has `num_nextn_predict_layers = 1`, so every draft token is another autoregressive pass through the same MTP head — the per-step overhead is paid k times.

### Why this matters

For latency-sensitive single-stream use, speculation is the main lever (at high concurrency the GPU is compute-saturated and speculation contributes almost nothing — see the 64-user column). Being capped at k=4 by framework overhead rather than by draft quality leaves measurable throughput on the table.

### Ask

Is fused multi-step draft decode feasible for these backends (`ROCM_AITER_MLA_SPARSE`, `DEEPSEEK_V32_INDEXER`, `KPOOL_TAIL`), or is there a structural reason the sparse-indexer metadata cannot be built once for all draft steps? If it is mainly a matter of hoisting the metadata build, I am happy to test a patch on this hardware — I have the harness and the baseline numbers above.

Related: #54359 (a correctness bug in the same kpool indexer path on ROCm), #53906 (GLM-5.3-Flash support).

### PR DESCRIPTION
**Pull request by Leoyzen**
## 🚀 Description

Fixes #54369 (CUDA side).

On CUDA, the sparse-MLA backends (`FlashMLASparse`, `DeepseekV32Indexer`) never declare
`supports_draft_decode_metadata_update`, so MTP draft decode falls back to rebuilding
attention metadata between every draft step (up to k=7 rebuilds per decode step at k=8,
per #54369). Under DCP, the indexer additionally could not participate at all: rank-local
sequence lengths must be re-derived from the speculator-advanced global lengths inside the
captured fused loop, and that path was not CUDA-graph-capture-safe.

This PR opts the two CUDA sparse backends into the fused framework from #46849:

- **FlashMLA sparse**: no-op update hook — the framework already advances positions,
  `seq_lens` and slot mappings in-graph, and the per-step topk indices are already
  rewritten in-forward on persistent buffers.
- **DSV3.2 indexer**: two-field in-place refresh — localized `decode.seq_lens`
  (recomputed via `get_dcp_local_seq_lens` from the advanced global lens, never an
  increment) and `deep_gemm` `schedule_metadata` (re-run into the same buffer slice; a
  stale captured schedule can deadlock the paged MQA logits kernel).
- **Capture-safety fix** in `get_dcp_local_seq_lens`: the rank offset was constructed as
  a CPU-scalar tensor on the caller's CUDA device, an illegal CPU→CUDA copy during graph
  capture (all TP workers died in `profile_cudagraph_memory`). It is now a Python scalar
  folded into the existing device arithmetic.
- **Eligibility guards**: the `advance_draft_positions=False` path no longer enables
  fused mode without declared backend support (previously a `NotImplementedError` crash
  at first propose); fused + MTP index sharing is rejected at startup for
  `num_mtp_layers > 1` (shared topk buffer aliasing).
- **Default-off opt-in**: `VLLM_ENABLE_FUSED_DRAFT_SPARSE_MLA` (with
  `VLLM_DISABLE_FUSED_DRAFT_SPARSE_MLA` override). No env vars set → behavior identical
  to pre-change.

## Known limitations

- Feature is opt-in (default off): on a compute-bound GLM-MTP TP8/DCP4 serving sweep the
  throughput delta was within the ±3.6% run-to-run noise band at concurrency 2–16
  (+12.7% at C1, −12% at saturated C32), so we do not flip the default in this PR.
- FlashInfer sparse variants (sm90/sm120) are audited capture-safe but not enabled here;
  a follow-up can flip them the same way.

## 🧮 Test plan / results

- [x] `.venv/bin/python -m pytest tests/v1/worker/test_gpu_autoregressive_speculator.py -q`
  → **38 passed** (12 new: DCP-localized refresh field equality, max-len clamp boundary
  propagation, env-gating matrix, index-sharing toggles, non-advancing fallback,
  capture/replay equality, producer-consumer ordering).
- [x] E2E, GLM-5.3 MTP, TP8 + DCP4, H200: feature enabled → server starts, FULL decode
  CUDA graphs capture cleanly (per-step-rebuild fallback warning gone); greedy
  acceptance counters identical to the pre-change baseline across 200-request runs × 3
  (8977/11319 accepted; one run bit-identical) — no acceptance drift.
- [x] Without the `get_dcp_local_seq_lens` fix, all 8 workers crash in
  `profile_cudagraph_memory` with "Cannot copy between CPU and CUDA tensors during CUDA
  graph capture" — reproduced and fixed.

## Related

- Fixes #54369 (CUDA + DCP side). The ROCm enablement is #52628 — complementary, no
  code overlap (different backends/arch); this PR covers FlashMLA-sparse + DSV3.2
  indexer + DCP-aware refresh, which no open PR addresses (checked before opening).
- Builds on the fused framework from #46849.

---

This PR was prepared with AI assistance (OpenCode / GLM-5.3). Every changed line was
reviewed and all listed tests were run by the submitter.

### CODE REVIEW

**Comment by Leoyzen** [comment]
### Perf data (GLM-5.3 MTP, TP8 + DCP4, H200, greedy)

Setup: same pod, same weights/config (Mooncake/KV-connector stripped for isolation),
FULL_AND_PIECEWISE CUDA graphs, K=3, 200 requests per run, greedy sampling. Acceptance
measured from spec-decode counters (`num_accepted_tokens_total / num_draft_tokens_total`)
per run. Baseline = current main behavior (per-step metadata rebuild, fallback warning
present); Fused = `VLLM_ENABLE_FUSED_DRAFT_SPARSE_MLA=1` (rebuild warning gone, fused
multi-step draft decode active).

**Correctness gate first**: acceptance is identical between the two arms — 79.3%
(8977/11319) with one run matching the baseline counters bit-for-bit, across 3 repeated
runs each. Greedy token streams match.

Throughput (tok/s, output):

| Concurrency | Baseline | Fused | Delta |
|---|---|---|---|
| 1 | 71.6 | 80.7 | **+12.7%** |
| 2 | 125.7 | 129.2 | +2.8% |
| 4 | 218.1 | 210.1 | −3.7% |
| 8 | 358.0 | 365.5 | +2.1% |
| 16 | 520.4 | 534.5 | +2.7% |
| 32 (saturated) | ~460 | ~404 | −12% |

Run-to-run noise band on this workload is ±3.6%, so:

- **C2–C16 is neutral** (within noise, slight positive trend).
- **C1 (latency-critical single stream) gains ~13%** — this is where removing the
  per-step host rebuild chain actually shows up end-to-end.
- **C32 regresses** — at saturation the GPU is already the bottleneck and the fused
  loop's extra in-graph refresh work (DCP-local seq-lens recompute + deep_gemm schedule
  rebuild every step) costs more than the host rebuild it replaces.

That's why this PR keeps the feature **opt-in (default off)**. For latency-sensitive /
low-concurrency serving it's a clear win; for saturated high-concurrency serving the
current per-step rebuild path remains the better default until the in-graph refresh is
further trimmed.

Separately from perf: the `get_dcp_local_seq_lens` capture-safety fix in this PR is a
hard prerequisite for *any* in-graph DCP metadata refresh — without it every TP worker
crashes during `profile_cudagraph_memory` ("Cannot copy between CPU and CUDA tensors
during CUDA graph capture"), which is how it was found.

**Comment by Leoyzen** [comment]
Added KPOOL_TAIL support (`KpoolTailMetadataBuilder.update_draft_decode_metadata`, commit 168bfd56b9): GLM-5.3-Flash (kpool=4) uses a dedicated storage-only backend for the circular tail cache, so without it the fused path fell back to per-step rebuilds on that model.

The hook recomputes the circular slot mapping (`own_block * kpool + pos % kpool`) in-graph on the same persistent `slot_mapping_buffer` — it must be a real refresh, not a no-op: a stale mapping silently corrupts draft-step tail-cache writes and collapses MTP acceptance. Gated on the same envs, auto-declines support when positions is None.

Tests: 5 new cases in `tests/v1/worker/test_gpu_autoregressive_speculator.py` (in-place refresh with unchanged data_ptr, pos%4 wrap, empty-batch early no-op, support decline without positions, env gating) — 43/43 in the file.

E2E A/B (GLM-5.3-Flash, 8xH200, TP8/EP8, fp8_ds_mla, FLASHMLA_SPARSE, MTP K=3, deterministic counting prompt, fused off = best of two rounds):

| concurrency | off (tok/s) | on (tok/s) | delta |
|---|---|---|---|
| 1 | 363.2 | 392.1 | +8.0% |
| 8 | 2336.8 | 2390.2 | +2.3% |
| 32 | 6325.4 | 6813.9 | +7.7% |
| 64 | 9760.0 | 9484.2 | ~flat |
| 128 | 10091.0 | 12914.5 | +28.0% |

MTP acceptance on natural English prompts unchanged (78% vs 85% baseline on a different mix). The KPOOL_TAIL hook was validated in this same deployment (no fallback warnings, `vllm:spec_decode` deltas healthy).
